File: Python/Best Time to Buy and Sell Stock III.py
# Say you have an array for which the ith element is the price of a given stock on day i.

# Design an algorithm to find the maximum profit. You may complete at most two transactions.

#not accepted
class Solution:
    # @param prices, a list of integer
    # @return an integer
	def maxProfit(self, prices):
		index = 0
		max1 = 0
		max2 = 0
		Max = 0
		for i in range(0,len(prices)):
			max1 = self.searchMax(prices,0,i)
			max2 = self.searchMax(prices,i+1,len(prices)-1)
			if Max < (max1+max2):
				Max = max1 + max2	
		return Max
	def searchMax(self,prices,start,end):
		Max = 0
		Min = 65536
		for i in range(start,end+1):
			if Min > prices[i]:
				Min = prices[i]
			diff = prices[i] - Min
			if diff > Max:
				Max = diff
		return Max
#accepted http://chaoren.is-programmer.com/posts/43727.html
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        length = len(prices)
        if length == 0: return 0
         
        maxProfitForward = []
        minPrice = prices[0]
        maxProfit = -1
        for currPrice in prices:
            minPrice = min(minPrice, currPrice)
            maxProfit = max(maxProfit, currPrice - minPrice)
            maxProfitForward.append(maxProfit)
         
        maxProfitBackward = []
        maxPrice = prices[-1]
        maxProfit  = -1
        for currPrice in reversed(prices):
            maxPrice = max(maxPrice, currPrice)
            maxProfit = max(maxProfit, maxPrice - currPrice)
            maxProfitBackward.insert(0, maxProfit)
         
        maxProfit = maxProfitForward[-1] # 0 or 1 transaction
        for i in xrange(length - 1): # 2 transactions
            maxProfit = max(maxProfit, maxProfitForward[i] + maxProfitBackward[i + 1])
        return maxProfit

File: Java/Best_Time_to_Buy_and_Sell_Stock_III.java
class Best_Time_to_Buy_and_Sell_Stock_III{
	public int maxProfit(int[] prices) {
        
    }
    //go through from left to right
    //find the max profit for a[0,i]
    //then the maxprofit for a[0,i+1] = max(a[0,i],a[i+1]-min);
    //similar process, scan from right to left
    //find max profit for a[i,a.length-1];
    //find the maxprofit of all combination of left and right
    private int linearSolution(int[] prices){
    	if (prices == null || prices.length == 0) {  
            return 0;  
        }  
        int n = prices.length;  
        int[] left = new int[n];  
        int[] right = new int[n];  
        int min = prices[0];  
        for (int i = 1; i < n; i++) {  
            left[i] = left[i - 1] > prices[i] - min ? left[i - 1] : prices[i] - min;  
            min = min < prices[i] ? min : prices[i];  
        }  
        int max = prices[n - 1];  
        for (int i = n - 2; i >= 0; i--) {  
            right[i] = right[i + 1] > max - prices[i] ? right[i + 1] : max - prices[i];  
            max = max > prices[i] ? max : prices[i];  
        }  
        int value = 0;  
        for (int i = 0; i < n; i++) {  
            value = value > left[i] + right[i] ? value : left[i] + right[i];  
        }  
        return value;   
    }
    private int bruteForceSolution(int[] prices){

    }
    private int oneTrans(int[] prices){
    	int max = Integer.MAX_VALUE;
    	int currMin = prices[0];
    	int index = 0;
    	for(int i=1;i<prices.length;i++){
    		int diff = prices[i] - currMin;
    		if(currMin > prices[i])
    			currMin = prices[i];
    		max = Math.max(max,diff);
    	}
    }
}